snap/validate.go

```diff
@@ -263,6 +264,44 @@ func validateAppSocket(socket *SocketInfo) error {
 	return validateSocketAddr(socket, "listen-stream", socket.ListenStream)
 }
 
+type appStartupOrder int
+
+const (
+	orderBefore appStartupOrder = iota
+	orderAfter
+)
+
+func validateAppStartup(app *AppInfo, order appStartupOrder) error {
+	orderStr := "before"
```

Comment:
When I looked at this I was wondering if we could make it a bit nicer with some helpers. Then I got carried away :) Sorry for this, I played around a bit and came up with the following code:  https://gist.github.com/mvo5/6e0ac963aacce9d0c78f1c7cf829f9c4

Please have a look, maybe you also like it. 

diff:
```
-func validateAppStartup(app *AppInfo, order appStartupOrder) error {
-       orderStr := "before"
-       deps := app.Before
-       if order == orderAfter {
-               orderStr = "after"
-               deps = app.After
+func (a appStartupOrder) String() string {
+       switch a {
+       case orderBefore:
+               return "before"
+       case orderAfter:
+               return "after"
+       }
+       panic(fmt.Sprintf("unsupported order: %v", a))
+}
+
+func appStartOrdering(app *AppInfo, order appStartupOrder) []string {
+       switch order {
+       case orderBefore:
+               return app.Before
+       case orderAfter:
+               return app.After
        }
+       panic(fmt.Sprintf("unsupported order: %v", order))
+}
 
-       for _, dep := range deps {
+func validateAppStartup(app *AppInfo, order appStartupOrder) error {
+       for _, dep := range appStartOrdering(app, order) {
                // dependency is not defined
-               if other, ok := app.Snap.Apps[dep]; !ok {
+               other, ok := app.Snap.Apps[dep]
+               if !ok {
                        return fmt.Errorf("cannot validate app %q startup ordering: needs to be started %s %q, but %q is not defined",
-                               app.Name, orderStr, dep, dep)
-               } else {
-                       // or it depends on us in the same order, eg. foo is
-                       // started after bar, but bar is to be started after foo
-                       // as well, validate only one dependency step
-                       otherDeps := other.Before
-                       if order == orderAfter {
-                               otherDeps = other.After
-                       }
-
-                       if strutil.ListContains(otherDeps, app.Name) {
-                               return fmt.Errorf("cannot validate app %q startup ordering: needs to be started %s %q, but %q is declared to start %s %q",
-                                       app.Name, orderStr, other.Name, other.Name, orderStr, app.Name)
-                       }
+                               app.Name, order.String(), dep, dep)
+               }
+               // it depends on us in the same order, eg. foo is
+               // started after bar, but bar is to be started after foo
+               // as well, validate only one dependency step
+               otherDeps := appStartOrdering(other, order)
+               if strutil.ListContains(otherDeps, app.Name) {
+                       return fmt.Errorf("cannot validate app %q startup ordering: needs to be started %s %q, but %q is declared to start %s %q",
+                               app.Name, order, other.Name, other.Name, order, app.Name)
                }
        }
        return nil
```

Reply:
Thanks for suggestion, looks nice indeed. I used the snippet you provided along with some minor tweaks.